src/clusterd/src/bin/clusterd.rs

```diff
@@ -195,38 +195,67 @@ async fn run(args: Args) -> Result<(), anyhow::Error> {
         )
     });
 
-    let secrets_reader = args
-        .secrets
-        .load()
-        .await
-        .context("loading secrets reader")?;
-    let config = mz_storage::Config {
-        build_info: &BUILD_INFO,
-        workers: args.workers,
-        timely_config,
-        metrics_registry,
+    let persist_clients = Arc::new(Mutex::new(PersistClientCache::new(
+        PersistConfig::new(&BUILD_INFO, SYSTEM_TIME.clone()),
+        &metrics_registry,
+    )));
+
+    // Start storage server.
+    let (_storage_server, storage_client) = mz_storage::serve(mz_storage::Config {
+        persist_clients: Arc::clone(&persist_clients),
+        workers: args.storage_workers,
+        timely_config: timely::Config {
+            worker: timely::WorkerConfig::default(),
+            communication: match args.storage_workers {
+                0 => bail!("--storage-workers must be greater than 0"),
+                1 => timely::CommunicationConfig::Thread,
+                _ => timely::CommunicationConfig::Process(args.storage_workers),
+            },
+        },
+        metrics_registry: metrics_registry.clone(),
         now: SYSTEM_TIME.clone(),
         connection_context: ConnectionContext::from_cli_args(
             &args.tracing.log_filter.inner,
             args.aws_external_id,
             secrets_reader,
         ),
-    };
-
-    // Initialize fail crate for failpoint support
-    let _failpoint_scenario = fail::FailScenario::setup();
-
-    let (_server, client) = mz_storage::serve(config)?;
+    })?;
+    info!(
+        "listening for storage controller connections on {}",
+        args.storage_controller_listen_addr
+    );
+    mz_ore::task::spawn(
+        || "storage_server",
+        GrpcServer::serve(
+            args.storage_controller_listen_addr,
+            BUILD_INFO.semver_version(),
+            storage_client,
+            ProtoStorageServer::new,
+        ),
+    );
 
+    // Start compute server.
+    let (_compute_server, compute_client) =
+        mz_compute::server::serve(mz_compute::server::Config {
+            metrics_registry,
+            persist_clients,
+        })?;
     info!(
-        "listening for controller connections on {}",
-        args.controller_listen_addr
+        "listening for compute controller connections on {}",
+        args.compute_controller_listen_addr
     );
-    GrpcServer::serve(
-        args.controller_listen_addr,
-        BUILD_INFO.semver_version(),
-        client,
-        ProtoStorageServer::new,
-    )
-    .await
+    mz_ore::task::spawn(
+        || "compute_server",
+        GrpcServer::serve(
+            args.compute_controller_listen_addr,
+            BUILD_INFO.semver_version(),
+            compute_client,
+            ProtoComputeServer::new,
+        ),
+    );
+
+    // TODO: unify storage and compute servers to use one timely cluster.
+
+    // Block forever.
+    future::pending().await
```

Comment:
i would just `select!` the 2 join handles here!

Reply:
🤷🏽 this matches what we do in `environmentd` (albeit in sync code), so just going to leave as is for now! Not opposed to cleaning it up in both places though.